src/hardware/vga_crtc.cpp

```diff
@@ -145,8 +145,9 @@ void vga_write_p3d5(io_port_t, io_val_t value, io_width_t)
 		if (IS_VGA_ARCH)
 			vga.config.line_compare=(vga.config.line_compare & 0x5ff)|(val&0x40)<<3;
 
-		if (IS_VGA_ARCH && (svgaCard==SVGA_None) && (vga.mode==M_EGA || vga.mode==M_VGA)) {
-			// in vgaonly mode we take special care of line repeats (excluding CGA modes)
+		if (IS_VGA_ARCH && (vga.mode == M_EGA || vga.mode == M_VGA)) {
+			// When using a VGA card in EGA and VGA modes, take
+			// special care to repeat lines (excluding CGA modes)
```

Comment:
I'm confused by this statement and/or the code. VGA cards display all 200-line modes line-doubled, so CGA screen modes on VGA cards are line-doubled as well, naturally. I think `vga.mode == M_CGA` should be added to the condition, no? I will test it anyway, but both the code and comment seem to be wrong.

Reply:
VGA cards line double modes according to two places in the CRTC.

One is a bit that says to doublescan and the vertical CRTC parameters have values that are half the actual raster output. This is typical of VGA hardware when INT 10h is asked to set up CGA/EGA 200-line graphics modes.

The other is the "max scanline" register which of course determines in graphics modes how many times the scanline repeats, and how many scanlines a character cell is in text mode. In 256-color 320x200 mode this register is usually "1" to say that every video memory scan line is rendered twice, unless the DOS game decides to program a 320x400 256-color mode by setting that register to zero.

Reply:
That's some good info @joncampbell123, but it doesn't answer my question why EGA is included in the conditional statement but CGA isn't.

We must double scan for CGA modes too on VGA adapters, not just for EGA modes.

Reply:
Found three touch points (two in vga_draw and one in CRTC) to enable double-scanning of CGA while using a VGA architecture, however I wasn't able to figure out where this slight wrapping or offset was happening.

![2023-04-15_11-13](https://user-images.githubusercontent.com/1557255/232246834-ab8935ae-57c8-4efe-8c5d-8e4180a047ac.png)

I'll leave the single commit + revert in the PR for future exploration (if someone search the commit log or reads about it here).

I suspect it would take more `if (VGA_ARCH)` conditions where the non-composite `M_CGA*` modes are handled (to differentiate the double vs single line handling, and possibly timing too).  Find them with: `git grep M_CGA | grep -v COMPOSITE`

Reply:
Odd and even lines are simply swapped on the above screenshot; zoom in and examine the top of the letters closely, then you can see it after some staring 😄 Hopefully that helps in getting to the bottom of it.

Reply:
I doubt the timings need changing. My understanding is we don't mess around with the standard timing values. As mentioned, seems like a simple odd/even line swapping bug with double scanning enabled.

Reply:
So CGA modes won't be double-scanned for now? That's fine, I just want to understand the current status, that's all 😄 If that's the case, we should also document that enabling double-scanning of CGA modes when a VGA card is emulated constitutes future work.

Reply:
Getting sometihng out that works for VGA and EGA and tweaking & improving it later. Now we know there's this gap left to be filled.

Reply:
I agree, I was just looking for confirmation on this because it wasn't 100% clear from the rest of the code.

Reply:
Included comment:

```c++
// TODO: Non-composite CGA modes should be included here too, as VGA
// cards also line-doubled those modes, however more work is needed to
// correctly draw double-scanned CGA lines.
}
```